offline-query.mdx

```diff
@@ -181,4 +183,72 @@ The user can specify
 [environment](/docs/resolver-environments),
 or [branch](/docs/branches)
 as parameters to `offline_query` in the same fashion
-as [online query](/docs/query-basics).
+as [online query](/docs/query-online).
+
+### Debugging offline queries
+
+Like [online queries](/docs/query-basics),
+offline queries can be debugged using the `explain` flag, which logs the query execution plan, and
+the `store_plan_stages` flag, which stores the outputs of each of the plan stages,
+which will be visible in the web dashboard's query detail view.
+More on these and other tips [here](/docs/debugging-queries).
+
+### Running large offline queries
+
+Offline queries run on a large number of rows require their own compute resources and infrastructure
+to ensure they complete and do not contend with resources from other workloads like the engine or branch server.
+Chalk handles this by launching a Kubernetes job that can runs asynchronously on dedicated compute nodes.
+To do so, set the `run_asynchronously` flag to `True` in the `offline_query` method.
+
+```python
+from chalk.client import ChalkClient
+
+client = ChalkClient()
+client.offline_query(
+  input={'user.id': range(1_000_000)},
+  output=['user.name'],
+  run_asynchronously=True,
+)
+```
+
+If you'd like to adjust the resource compute or memory for the job,
+set the `resources` parameter to a custom [`ResourceRequests`](/api-docs#ResourceRequests) object.
+Ensure that the request values are compatible with the resources available on Kubernetes cluster.
+
+Another option to consider is to parallelize the computation of the offline query by dividing the input into shards.
+If the `num_shards` parameter is set,
+Chalk will split the input into `num_shards` shards and run each shard in a separate pod.
+The `num_workers` parameter sets the maximum number of pods that can run concurrently.
+If the number of shards is large compared to the size of the Kubernetes cluster,
+set the `num_workers` parameter to a smaller number to ensure that Chalk can launch enough pods to run the query.
+
+```python
+from chalk.client import ChalkClient
+
+client = ChalkClient()
+client.offline_query(
+  input={'user.id': range(1_000_000)},
+  output=['user.name'],
+  run_asynchronously=True,
+  num_shards=50,
+  num_workers=10,
+)
+```
+
+For the above example, Chalk will split the input spine into 50 shards of 20,000 rows each.
+It will then launch jobs for each shard, while not exceeding 10 concurrent pods at once.
+
+If you lose track of the state of a long-running offline query, you'll always be able to retrieve the outputs later.
```

Comment:
? is this like...notebook dying?

Reply:
lmao. the notebook shows code that wouldn't work anymore